New page in quickstart derived types (#98)
fortran-lang/fortran-lang.org: learn/quickstart/derived_types.md

---
layout: book
title: Derived types
permalink: /learn/quickstart/derived_types
---

As discussed previously in [Variables]({{site.baseurl}}/learn/quickstart/variables) there are five built-in data types in Fortran. _Derived types_ is a special form of a data type that can encapsulate other built-in types as well as other derived types. It could be considered as the equivalent of _struct_ in the C and C++ programming languages.

## A quick take on derived types

Here's an example of a basic derived type:

```fortran
type :: t_pair
    integer :: i
    real    :: x
end type
```

The syntax to create a variable of type _t_pair_ and access its members is:
```fortran
! declare
type(t_pair) :: pair
! initialize
pair%i   = 1
pair%x   = 0.5
```

{% include note.html content="The percentage symbol `%` is used to access the members of a derived type." %}

In the above snippet, we declared an instance of a derived type and initialized its members explicitly.

> You can also initialize derived type members by invoking the derived type constructor:

Reviewer: ```suggestion
You can also initialize derived type members by invoking the derived type constructor.
```

{% include note.html content="If the `attribute: bind(c)` or the `statement: sequence` is used then a derived type cannot have the `attribute: extends` and visa-versa." %}

The `sequence` attribute may be used only to declare that the following  members should be accessed in the same order as they are defined within the derived type. 

Example with `sequence`:
```fortran
type :: foo
sequence
integer :: var1
real    :: var2
end type
! init
type(foo) :: bar
bar = foo(1, 0.5)
```
{% include note.html content="The use of statement `sequence` presupposes that the data types defined below are neither of `allocatable` nor of `pointer` type. Furthermore, it does not imply that these data types  will be stored in memory in any particular form, there is no relation to `contigeous` attribute." %}

The _access-type_ attributes `public` and `private` if used, declare that all [member-variables] declared below will be automatically assigned the attribute accordingly. 

The attribute `bind(c)` is used to achieve compatibility between Fortran's derived type and C's struct.

Example with `bind(c)`:
```fortran

> module mymod

Reviewer 2: I would remove this line, or a `end module mymod` should be added in the example

> use iso_c_bindings

Reviewer 2: ```suggestion
use iso_c_bindings, only: c_int
```

Reviewer 3: also `iso_c_binding` without an `s` at the end

implicit none
type, bind(c) :: mytype
    integer(c_int) :: i
end type
```
matches the following C struct:
```c
struct{
    int i
}mytype;
```
{% include note.html content="A fortran derived type with the attribute `bind(c)` cannot have the `sequence` and `extends` attributes. Furthermore it cannot contain any Fortran `pointer` or `allocatable` types." %}

`parameterized-decleration-list`: is an optional feautre. If used, then the parameters must be listed in place of [parameterized-definition-statements] and must be either `len` or `kind` parameters or both. 

Example of a derived type with `parameterized-decleration-list` and with the `attribute: public`:
 ```fortran
module mymod
implicit none
private

type, public :: matrix(rows, cols, k)
  integer, len :: rows, cols
  integer, kind :: k = kind(0.0)
  real(kind = k), dimension(rows, cols) :: values
end type matrix

> end module

Reviewer 2: ```suggestion
end module mymod
```

program test
use mymod
implicit none
type(matrix(rows=5, cols=5)) :: m
end program
 ```

> {% include note.html content="In this example the parameter **k** has already been assigned a default value of `kind(0.0)`, that is of floating point single precision, for that reason it can be ommited, as it is the case here in the declaration inside the main program." %}

Reviewer 2: ```suggestion
{% include note.html content="In this example the parameter **k** has already been assigned a default value of `kind(0.0)`, that is of floating point single precision. Therefore, it can be omitted, as it is the case here in the declaration inside the main program." %}
```

The attribute `extends` was added in F2003 standard and introduces an important feature of the object oriented paradigm (OOP), namely the inheritance. It allows code reusability by letting children-derived-types like this: `type, extends(parent) :: child` to inherit all the members and functionality from a parent-derived-type: `type :: parent`.  

Example with the attribute `extends`: 
```fortran
module mymod
implicit none
private
public t_date, t_address, t_person, t_employ ! note another way of using the public attribute by gathering all public data types in one place

type :: t_date
    integer                         :: year, month, day
end type

type :: t_address
    character(len=:), allocatable   :: city, road_name
    integer                         :: house_number
end type

type, extends(t_address) :: t_person
    character(len=:), allocatable   :: first_name, last_name, e_mail  
end type

type, extends(t_person)  :: t_employ
    type(t_date)                    :: hired_date
    character(len=:), allocatable   :: position
    real                            :: monthly_salary
end type

> end module

Reviewer 2: ```suggestion
end module mymod
```

program test
use mymod
implicit none
type(t_employ) :: employ

! initialization
employ%hired_date%year  = 2020 ! t_employ has access to type(t_date) members not because of extends but because a type(t_date) was declared within t_employ
employ%hired_date%month = 1
employ%hired_date%day   = 20
employ%first_name       = 'Johny' !t_employ has acces to t_person, and inherits its members due to extends 
employ%last_name        = 'Doe'
employ%city             = 'London' ! t_employ has access to t_address, because it inherits from t_person, that in return inherits from t_address
employ%road_name        = 'BigBen'
employ%house_number     = 1
employ%position         = 'Intern'
employ%monthly_salary   = 0.0

> end program

Reviewer 2: ```suggestion
end program test
```

    !1st case: simple built-in type with access attribute and [init]
    integer, private :: i = 0 ! private hides it from use outside of the t_example's scope. The default initialization [=0] is the [init] part. 

    !2nd case: dynamic 1d_array
    real, allocatable, dimension(:) :: x
    ! the same as
    real, allocatable :: x(:) ! parenthesis implies dimension(:) and is one of the possible [attr-dependent-spec].

    !3rd case: protected
    integer, protected :: i ! In contrary to private, protected allows access to i assigned value outside of t_example but is not definable, i.e. a value may be assigned to i only within t_example.

    !4th case: pointer, with [init]
    real, pointer :: x(:) => null() ! the [init] part is the [=>null()], pointers are discussed in the Advanced programming mini-book. 

    !5th case: coarray
    real, allocatable, codimension[:] :: z(:) ! a 1d_dynamic array shared in all threads. Coarrays will be discussed in the Advanced programming mini-book.
    !or 
    real, allocatable :: z(:)[:] ! here the [:] is the [attr-dependent-spec] and implies the codimension[:].

    !6th case: contiguous
    real, contiguous, pointer :: x(:)
end type
```

> {% include note.html content="In the above example the cases 4, 5 and 6 make use of Fortran `pointer` and `coarray` features that have not been addressed in this quickstart tutorial. However, they are presented here, in order for the readers to know that these features do exist and be able to recognize them. These features will be covered in detail in the upcoming `Advanced programing` mini-book." %}

Reviewer: With Manning I learned a useful principle: "Don't introduce it until you have to use it". This only puts unnecessary burden on the reader. I suggest we move this section to the advanced tutorial.

{% include note.html content="It is impossible to describe type-bound procedures in their full syntax without delving into OOP features of modern Fortran. For that reason only a simple example is provided in this final part, to demonstrate a very basic use." %}

Example of a derived type with basic bound-procedure:

```fortran
module mymod
implicit none
private
public t_square

type :: t_square
    real :: side
    contains
        procedure :: area !procedure declaration
end type

contains
    ! procedure definition
    real function area(self) result(res)
        class(t_square), intent(in) :: self 
        res = self%side * self%side
    end function
end module

program main
use mymod
implicit none
! variables declaration
type(t_square) :: sq
real :: x, side

! variables initialization
side    = 0.5
sq%side = side

x       = sq%area() ! self does not appear here, it has been passed implicitly
! do stuff with x...
end program
```
What is new:

>  - **self** is a random name that was chosen to represent the derived type t_square that is passed as an argument to the bound-function in order to have access to its data-members. By passing it like that it is ensured that later during its use the t_square will be passed automatically and not by the client.

Reviewer: ```suggestion
 - **self** is an arbitrary name that we chose to represent the instance of the derived type t_square inside the type-bound function. This allows us to access its members and to automatically pass it as an argument when we invoke a type-bound procedure.
```

>  - in order to have the above functionality the new keyword `class` replaced the `type` one. With `class` the OOP feature *polymorphism* is introduced. 

Reviewer: ```suggestion
 - We now use `class(t_square)` instead of `type(t_square)` in the interface of the `area` function. This allows us to invoke the `area` function with any derived type that extends `t_square`.
```

Author: Hi @milancurcic, I will accept your suggestion. What do you think about leaving the last sentence like:
"With the keyword, `class` the OOP feature *polymorphism* is introduced.", in order, in future, when a complete OOP tutorial has been uploaded the word *polymorphism* to be transformed to hyperlink that redirects to the corresponding page.

Reviewer: Yes, sounds good. You can write it without comma, just 

"With the keyword `class` the OOP feature _polymorphism_ is introduced." 

And even better, in active voice:

"The keyword `class` introduces the OOP feature, _polymorphism_."

>  - since the bound-procedure **area** was defined as a function it cannot be called by itself, it can only appear as **rhs** object, that is why it is used like `x = sq%area()`. The 'stand-alone' functionality is covered by a subroutine, and the above example should be modified like:

Reviewer: The 3rd bullet actually didn't belong in the "what's new" list, but is its own separate lesson. So I put it in a new paragraph.

```suggestion

In the above example, the type-bound procedure **area** is defined as a function and can be invoked only in an expression, for example `x = sq%area()` or `print *, sq%area()`. If you define it instead as a subroutine, you can invoke it from its own `call` statement:
```

 ```fortran
 ! change within module
 contains
    subroutine area(self, x)
        class(t_square), intent(in)     :: self

>         real,            intent(in out) :: x

Reviewer: ```suggestion
        real,            intent(out) :: x
```

>         x = self%side * self%side

Reviewer: ```suggestion
        x = self%side**2
```

    end subroutine

! change within main program
call sq%area(x)
! do stuff with x...
 ```

>  In this case there are two arguments in definition, one similar as before the **self** of type `t_square` and the second one a real variable **x** that should be assigned the calculated area and returned back for further use. 

Reviewer: ```suggestion
In contrast to the example with the type-bound function, we now have two arguments: 

* `class(t_square), intent(in) :: self`, which is the instance of the derived type itself
* `real, intent(out) :: x` which is used to store the calculated area into and return to the caller. 
```
